Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs

```diff
@@ -0,0 +1,252 @@
+                CredentialEnvironmentVariable,
+                HelpPermalink));
+        }
+
+        /// <summary>Unsynchronized helper method to obtain the Application Default Credentials.</summary>
+        private GoogleCredential GetDefaultCredentialUnsynchronized()
+        {
+            string credentialPath = null;
+            
+            // 1. First try the environment variable.
+            try
+            {
+                credentialPath = GetEnvironmentVariable(CredentialEnvironmentVariable);
+                if (!String.IsNullOrWhiteSpace(credentialPath))
+                {
+                    return CreateDefaultCredentialFromFile(credentialPath);
+                }
+            }
+            catch (Exception e)
+            {
+                // Catching generic exception type because any corrupted file could manifest in different ways including 
+                // but not limited to the System, System.IO or from the Newtonsoft.Json namespace.
+
+                throw new InvalidOperationException(String.Format("Error reading credential file from location {0}: {1}"
+                                        + "\nPlease check the value of the Environment Variable {2}",
+                                        credentialPath,
+                                        e.Message,
+                                        CredentialEnvironmentVariable));
+            }
+
+            // 2. Then try the well known file.
+            try
+            {
+                credentialPath = GetWellKnownCredentialFilePath();
+                if (!String.IsNullOrWhiteSpace(credentialPath))
+                {
+                    return CreateDefaultCredentialFromFile(credentialPath);
+                }
+            }
+            catch (FileNotFoundException)
+            {
+                // File is not present, eat the exception and move on to the next check.
+            }
+            catch (IOException e)
+            {
+                throw new InvalidOperationException(String.Format("Error reading credential file from location {0}: {1}"
+                                  + "\nPlease rerun 'gcloud auth login' to regenerate credentials file.",
+                                  credentialPath,
+                                  e.Message));
+            }
+
+            // 3. Then try the compute engine, if not already checked.         
+            if (!checkedComputeEngine)
+            {
+                Logger.Debug("Checking whether the application is running on ComputeEngine.");
+                bool isRunningOnComputeEngine = ComputeCredential.IsRunningOnComputeEngine().Result;
+                checkedComputeEngine = true;
+                if (isRunningOnComputeEngine)
+                {
+                    Logger.Debug("ComputeEngine check passed. Loading ComputeEngine Credentials.");
+                    return GoogleCredential.FromCredential(new ComputeCredential());
```

Comment:
Seems weird to me. How does it help to create an empty object and then send it to GoogleCredential. Maybe I need more context, but it seems weird.. What do you think?